src/celeritas/optical/interactor/MieInteractor.hh

```diff
@@ -0,0 +1,151 @@
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+CELER_FUNCTION
+MieInteractor::MieInteractor(NativeCRef<MieData> const& shared,
+                             ParticleTrackView const& particle,
+                             Real3 const& direction,
+                             OptMatId const& mat_id)
+    : inc_dir_(direction)
+    , inc_pol_(particle.polarization())
+    , mie_params_(shared.mie_record[mat_id])
+{
+    CELER_EXPECT(shared);
+    CELER_EXPECT(mat_id < shared.mie_record.size());
+    CELER_EXPECT(is_soft_unit_vector(inc_dir_));
+    CELER_EXPECT(is_soft_unit_vector(inc_pol_));
+    CELER_EXPECT(soft_zero(dot_product(inc_dir_, inc_pol_)));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample a single optical Mie scattering event.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MieInteractor::operator()(Engine& rng) const
+{
+    Interaction result;
+    Real3& new_dir = result.direction;
+    Real3& new_pol = result.polarization;
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+
+    UniformRealDist sample_phi(0, real_type(2 * constants::pi));
+    UniformRealDist sample_r(0, 1);
+    UniformRealDist sample_g(0, 1);
+
+    real_type u = sample_g(rng);
+    real_type r = sample_r(rng);
+
+    // Check if scattering should be forward scattering or backward scattering
+    bool is_forward = u < mie_params_.forward_ratio;
+    real_type g = (is_forward ? mie_params_.forward_g : mie_params_.backward_g);
+
+    // Sample cos theta
+    real_type costheta
+        = (g != 0) ? (2. * r * (1. + g) * (1. + g) * (1. - g + g * r)
+                          / ((1. - g + 2. * g * r) * (1. - g + 2. * g * r))
+                      - 1.)  // will make it cleaner
+                   : (2 * r - 1);
+
+    // Reverse cos theta if backward_g is chosen
+    if (!is_forward)
+        costheta = -costheta;
+
+    real_type phi = sample_phi(rng);
+
+    // Creating new direction
+    new_dir = from_spherical(costheta, phi);
+
+    // Build polarization vector
+    SoftZero const soft_zero{SoftEqual{}.rel()};
+    do
+    {
+        do
+        {
+            new_pol = make_unit_vector(make_orthogonal(inc_pol_, new_dir));
+
+            // Reject rare case of polarization and new direction being
+            // coincident leading to loss of orthogonality
+
+        } while (CELER_UNLIKELY(!soft_zero(dot_product(new_pol, new_dir))));
+
+        if (!BernoulliDistribution{0.5}(rng))
+        {
+            // Flip direction with 50% probability: there are two polarizations
+            // perpendicular to the new direction and the original polarization
+            new_pol = -new_pol;
+        }
+        // Accept with the probability of the scattered polarization overlap
+        // squared
+    } while (RejectionSampler{ipow<2>(
+        clamp<real_type>(dot_product(new_pol, inc_pol_), -1, 1))}(rng));
```

Comment:
This looks like a copy-paste error from Rayleigh; I don't think Mie has the same $`\sin^2 \theta`$ rejection.